src/Security/TOTPManager.php

```diff
@@ -0,0 +1,495 @@
+     * @throws \JsonException
+     */
+    public function verifyBackupCodeForUser(string $code, int $users_id, bool $consume_code = true): bool
+    {
+        global $DB;
+
+        $tfa = $DB->request([
+            'SELECT' => ['2fa'],
+            'FROM' => 'glpi_users',
+            'WHERE' => [
+                'id' => $users_id
+            ],
+            'LIMIT' => 1
+        ])->current()['2fa'] ?? null;
+        if ($tfa === null) {
+            return false;
+        }
+        $tfa = json_decode($tfa, true, 512, JSON_THROW_ON_ERROR);
+        if (!isset($tfa['backup_codes'])) {
+            return false;
+        }
+        foreach ($tfa['backup_codes'] as $i => $hash) {
+            if (password_verify($code, $hash)) {
+                if ($consume_code) {
+                    unset($tfa['backup_codes'][$i]);
+                    $DB->update('glpi_users', [
+                        '2fa' => json_encode($tfa, JSON_THROW_ON_ERROR),
+                    ], [
+                        'id' => $users_id
+                    ]);
+                }
+                return true;
+            }
+        }
+        return false;
+    }
+
+    /**
+     * Get the 2FA enforcement level for a user
+     * @param int $users_id ID of the user
+     * @return int One of the ENFORCEMENT_* constants
+     * @phpstan-return self::ENFORCEMENT_*
+     * @throws \Exception
+     */
+    public function get2FAEnforcement(int $users_id): int
+    {
+        global $CFG_GLPI, $DB;
+
+        $user_optional = $DB->request([
+            'SELECT' => ['2fa_unenforced'],
+            'FROM' => 'glpi_users',
+            'WHERE' => [
+                'id' => $users_id
+            ],
+            'LIMIT' => 1
+        ])->current()['2fa_unenforced'] ?? 0;
+
+        if ($user_optional) {
+            return self::ENFORCEMENT_OPTIONAL;
+        }
+
+        $enforced = false;
+        if ((int) $CFG_GLPI['2fa_enforced'] === 1) {
+            $enforced = true;
+        } else {
+            // Check entity configuration
+            $entities = PermissionManager::getInstance()->getAllEntities($users_id);
+            foreach ($entities as $entity) {
+                if (\Entity::getUsedConfig('2fa_enforcement_strategy', $entity, '', true) === 1) {
+                    $enforced = true;
+                    break;
+                }
+            }
+            if (!$enforced) {
+                // Check profile configuration
+                $profiles = \Profile_User::getUserProfiles($users_id);
+                if (!empty($profiles)) {
+                    $enforced = $DB->request([
+                        'SELECT' => ['2fa_enforced'],
+                        'FROM' => 'glpi_profiles',
+                        'WHERE' => [
+                            'id' => $profiles,
+                            '2fa_enforced' => 1
+                        ]
+                    ])->count() > 0;
+                }
```

Comment:
This implies that a user that has a limited profile, that does not enforce 2FA, as its default profile and an admin profile, that enforces 2FA, and that is used really often, then the user will be forced to use 2FA on both profiles. Is it possible to activate the 2FA only on profile selection?

Reply:
We discussed about it with @orthagh.

Requesting a OTP code on login as long as user has access to an entity that enforces 2FA or is part of a group/profile that enforces 2FA is the expected behaviour.

The only thing to add it a :information_source: block on corresponding configuration forms to clearly explains this.